runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ParDoBoundMultiTranslator.java

```diff
@@ -314,8 +314,12 @@ private static <InT, OutT> void doTranslatePortable(
 
     final RunnerApi.PCollection input = pipeline.getComponents().getPcollectionsOrThrow(inputId);
     final PCollection.IsBounded isBounded = SamzaPipelineTranslatorUtils.isBounded(input);
+
+    // No key coder required if it is a stateless ParDo or a stateful ParDo without user timers, see
+    // SamzaStateRequestHandlers. The key coder is required for instantiating
+    // SamzaTimerInternalsFactory, see DoFnOp
     final Coder<?> keyCoder =
-        StateUtils.isStateful(stagePayload)
+        stagePayload.getTimersCount() > 0
```

Comment:
This might be problematic since there might be cases that we have state but no timers. So we still need to figure out the keyCoder for that case.

Reply:
@xinyuiscool My thought was the same as you that we need to parse the key coder for the Pardo with user states, but I realized that the key coder information is not in use for instantiating user state APIs.

In the current implementation, the portable mode always uses the fixed key coder `ByteStringCoder` for all keys:
```java
    final SamzaStoreStateInternals.Factory<ByteString> stateInternalsFactory =
        SamzaStoreStateInternals.createStateInternalsFactory(
            transformId, ByteStringCoder.of(), context, pipelineOptions, executableStage);
```
https://github.com/apache/beam/blob/master/runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStateRequestHandlers.java#L107

I have verified this change by using below test job as well and the pipeline worked well:
```java
public class MultiParDosWithUserStateTest {

  public static void main(String[] args) {
    SamzaPipelineOptions pipelineOptions = SamzaPortableExampleOptionsFactory.create(args);
    Pipeline pipeline = Pipeline.create(pipelineOptions);

    pipeline
        .apply(
            "Read From Kafka",
            LiKafkaIO.<PageViewEvent>read()
                .withTopic("PageViewEvent")
                .withCluster(TRACKING)
                .withoutMetadata())
        .apply(Values.create())
        .apply(
            WithKeys.of((PageViewEvent pv) -> String.valueOf(pv.header.memberId))
                .withKeyType(TypeDescriptors.strings()))
        .apply(
            "First stateful ParDo",
            ParDo.of(
                new DoFn<KV<String, PageViewEvent>, KV<Integer, PageViewEvent>>() {

                  @StateId("first-state-id")
                  private final StateSpec<ValueState<Integer>> spec = StateSpecs.value();

                  @ProcessElement
                  public void process(
                      ProcessContext context,
                      BoundedWindow window,
                      @StateId("first-state-id") ValueState<Integer> count) {
                    int currentValue = MoreObjects.firstNonNull(count.read(), 0);
                    System.out.println(
                        "first-state-id: "
                            + currentValue
                            + ", "
                            + Thread.currentThread().getName());
                    count.write(currentValue + 1);

                    context.output(KV.of(123, context.element().getValue()));
                  }
                }))
        .apply(
            "Second stateful ParDo",
            ParDo.of(
                new DoFn<KV<Integer, PageViewEvent>, KV<String, PageViewEvent>>() {
                  @StateId("second-state-id")
                  private final StateSpec<ValueState<Integer>> spec = StateSpecs.value();

                  @ProcessElement
                  public void process(
                      ProcessContext context,
                      BoundedWindow window,
                      @StateId("second-state-id") ValueState<Integer> count) {
                    int currentValue = MoreObjects.firstNonNull(count.read(), 0);
                    System.out.println(
                        "second-state-id: "
                            + currentValue
                            + ", "
                            + Thread.currentThread().getName());
                    count.write(currentValue + 1);

                    KV<Integer, PageViewEvent> element = context.element();
                    context.output(KV.of("abc", element.getValue()));
                  }
                }))
        .apply(ConsoleOutput.of(kv -> kv.getKey() + ", " + kv.getValue()));

    pipeline.run();
  }
}
```